### ISSUE TITLE
Wrong batch derivation semantics related to sequencing window

### ISSUE THREAD

**Issue by LEAFERx** [a16z/magi#213]
# Problem

In current batches derivation stage, when seq window of an epoch `E` passed, it will derive empty batches belonging to `E` and *1 empty batch* at the beginning of epoch `E+1` to advance the `engine_driver.safe_epoch`.

https://github.com/a16z/magi/blob/82e2c51fb7a62c48d505d222b11d1ff5f784d129/src/derive/stages/batches.rs#L143-L157

However, at the time when the first empty batch of epoch `E+1` is derived from the stage, the seq window of `E+1` might not pass yet.
It will cause the chain to fork if the sequencer missed the seq window of epoch `E` but managed to sequence a non-empty batch in time in epoch `E+1`.

# Suggest Solution

Only advance `engine_driver.safe_epoch` to `E+1` and create empty batches of epoch `E+1` when seq window of `E+1` is passed.

```rust
let epoch = if next_timestamp < next_epoch.timestamp {
   // We're deriving empty batches in the current epoch
  Some(epoch)
} else if current_l1_block > next_epoch.number + seq_window_size {
  // The `next_epoch` also passed seq window, so it is safe to create empty batches in `next_epoch` and advance to it
  Some(next_epoch)
} else {
  // The `next_epoch` is still within its seq window, so don't create empty batches
  None
};

epoch.map(|epoch| BlockInput {
  epoch: epoch.number,
  timestamp: next_timestamp,
  transactions: Vec::new(),
  l1_inclusion_block: current_l1_block,
})
```

**Comment by ncitron** [comment]
Yep this looks right. Good catch! Would love it you can open a PR to adjust.

**Comment by ncitron** [comment]
I think this is related to at #207 as well. This should resolve the issue right?
